### DISCUSSION TITLE
does fastify hate middleware that much? Need Help

### DISCUSSION THREAD

**Discussion by ayami123** [fastify/fastify#5428] [Q&A]
I have a code which manipulate the request.body or request.query of the parameters

```javascript
app.register(require("fastify-raw-body"),{
        field: "rawBody", // change the default request.rawBody property name
        global: true, // add the rawBody to every request. Default true
        encoding: "utf8", // set it to false to set rawBody as a Buffer. Default utf8
        runFirst: true, // get the body before any preParsing hook change/uncompress it. Default false
        routes: ["/api/sample"], // specify the routes where you want to apply the plugin
      });


 fastify.post("/login", {
        onRequest: [
            paramsParser(["username", "password"], []),
        ],
    }, async (request, reply) => {
        crud.login(request, reply);
    });
```

it's so annoying that it's sending unidentified in request.body
after some researching
I found out that you are sending streams
so I tried to create another middleware which turn those streams into body 
which again
breaks the code
as fastify needs to parse that stream at the end

so this code is also useless
```javascript
 const rawDataChunks = [];

        request.raw.on('data', (chunk) => {
            rawDataChunks.push(chunk);
        });
    
        request.raw.on('end', () => {
            request.body = JSON.parse(Buffer.concat(rawDataChunks).toString());
            done();
        });
```

I'm out of options

Need help

**Comment by metcoder95** [comment]
Why not use the `preValidator` hook instead?
The body is not parsed until before the `onPreValidation` is called, so you can intercept it accordingly. It will depend heavily on your `content-type` header of the request to hint `fastify` how to parse it.

https://fastify.dev/docs/latest/Reference/Hooks/#prevalidation
